MediaBrowser.Controller/MediaEncoding/TranscodingSegmentCleaner.cs

```diff
@@ -0,0 +1,188 @@
+    {
+        _job = job;
+        _logger = logger;
+        _config = config;
+        _fileSystem = fileSystem;
+        _mediaEncoder = mediaEncoder;
+        _segmentLength = segmentLength;
+    }
+
+    /// <summary>
+    /// Start timer.
+    /// </summary>
+    public void Start()
+    {
+        _timer = new Timer(TimerCallback, null, 20000, 20000);
+    }
+
+    /// <summary>
+    /// Stop cleaner.
+    /// </summary>
+    public void Stop()
+    {
+        DisposeTimer();
+    }
+
+    /// <summary>
+    /// Dispose cleaner.
+    /// </summary>
+    public void Dispose()
+    {
+        Dispose(true);
+        GC.SuppressFinalize(this);
+    }
+
+    /// <summary>
+    /// Dispose cleaner.
+    /// </summary>
+    /// <param name="disposing">Disposing.</param>
+    protected virtual void Dispose(bool disposing)
+    {
+        if (disposing)
+        {
+            DisposeTimer();
+        }
+    }
+
+    private EncodingOptions GetOptions()
+    {
+        return _config.GetEncodingOptions();
+    }
+
+    private async void TimerCallback(object? state)
+    {
+        if (_job.HasExited)
+        {
+            DisposeTimer();
+            return;
+        }
+
+        var options = GetOptions();
+        var enableSegmentDeletion = options.EnableSegmentDeletion;
+        var segmentKeepSeconds = Math.Max(options.SegmentKeepSeconds, 20);
+
+        if (enableSegmentDeletion)
+        {
+            var downloadPositionTicks = _job.DownloadPositionTicks ?? 0;
+            var downloadPositionSeconds = Convert.ToInt64(TimeSpan.FromTicks(downloadPositionTicks).TotalSeconds);
+
+            if (downloadPositionSeconds > 0 && segmentKeepSeconds > 0 && downloadPositionSeconds > segmentKeepSeconds)
+            {
+                var idxMaxToRemove = (downloadPositionSeconds - segmentKeepSeconds) / _segmentLength;
+
+                if (idxMaxToRemove > 0)
+                {
+                    await DeleteSegmentFiles(_job, 0, idxMaxToRemove, 0, 1500).ConfigureAwait(false);
+                }
+            }
+        }
+    }
+
+    private async Task DeleteSegmentFiles(TranscodingJob job, long idxMin, long idxMax, int retryCount, int delayMs)
+    {
+        if (retryCount >= 10)
+        {
+            return;
+        }
+
+        var path = job.Path ?? throw new ArgumentException("Path can't be null.");
+
+        _logger.LogDebug("Deleting segment file(s) index {Min} to {Max} from {Path}", idxMin, idxMax, path);
+
+        await Task.Delay(delayMs).ConfigureAwait(false);
+
+        try
+        {
+            if (job.Type == TranscodingJobType.Hls)
+            {
+                DeleteHlsSegmentFiles(path, idxMin, idxMax);
+            }
+        }
+        catch (IOException ex)
+        {
+            _logger.LogError(ex, "Error deleting segment file(s) {Path}", path);
+
+            await DeleteSegmentFiles(job, idxMin, idxMax, retryCount + 1, 500).ConfigureAwait(false);
```

Comment:
If this throws an exception, and a file does not get deleted it will be enumerated the next timer tick and will fail again

Reply:
Any suggestions? The only reason I can think of for this to fail is that the user has set incorrect permissions for the transcodes folder.

Reply:
Yes but its the filesystem so stuffs bound to happen.
I mean as this whole class is already scoped just remember all segments you already tried to delete in a list and ignore them on any subsequent loops.

Reply:
I don't see a scenario where one segment deletion failing due to non-transient errors (transient would be ok to retry) doesn't mean all future deletions for any segment won't fail as well. Wouldn't it make more sense to look at the IO error and simply stop the cleaner if it's a perm-related error?

Reply:
If it's a perm-related issue then the transcoding won't happen at all, or it should break immediately (due to ffmpeg failing).

Reply:
Then I'm not sure I see the point of keeping track of failures. Either it can be retried later or none of the deletions will ever succeed. Or am I missing something?

Reply:
It may be locked by system programs, which is more likely to happen on Windows, such as virus scanning. If the system does not unlock some segment files, Jellyfin will continue to throw IOException and record it in the log. I'm okay with both since it's more of a cosmetic change to the log.

@JPVenson Any more comments on this?

Reply:
Yea the logbloat is my biggest concern. If you would just not log the subsequent errors i am fine with it.

Reply:
I mean, if we want to make it sofisticated you could use a FileSystemWatcher and a Channel with a delay. So you will only ever process an item once without a loop.

Reply:
That one seems a bit overkill. Maybe lowering the log level to debug is also feasible? This should avoid spamming in the log file in general use cases.

Reply:
I drop the excludeFilePaths and lower the log level to debug to avoid spamming in the log file. Now even if individual segments fail to be deleted, they will not be recorded in the normal log unless the user adjusts the log level to debug.